### PR TITLE
Fix fields merging and improve performance

### PR DESCRIPTION
**Pull request by ghostdogpr**
Fields were not properly merged during execution. I moved the merging to the validation phase and added some performance optimizations.

Before:
```
[info] Benchmark                               Mode  Cnt      Score      Error  Units
[info] GraphQLBenchmarks.fragmentsCaliban     thrpt    5     27.739 ±    2.741  ops/s
[info] GraphQLBenchmarks.fragmentsSangriaNew  thrpt    5     24.269 ±    4.216  ops/s
[info] GraphQLBenchmarks.fragmentsSangriaOld  thrpt    5      0.165 ±    0.011  ops/s
[info] GraphQLBenchmarks.introspectCaliban    thrpt    5   1585.284 ±  183.560  ops/s
[info] GraphQLBenchmarks.introspectSangria    thrpt    5    499.669 ±   19.512  ops/s
[info] GraphQLBenchmarks.simpleCaliban        thrpt    5  24539.809 ± 1188.386  ops/s
[info] GraphQLBenchmarks.simpleSangria        thrpt    5   7551.214 ± 1423.378  ops/s
```
After:
```
[info] Benchmark                               Mode  Cnt      Score     Error  Units
[info] GraphQLBenchmarks.fragmentsCaliban     thrpt    5     44.638 ±   1.523  ops/s
[info] GraphQLBenchmarks.fragmentsSangriaNew  thrpt    5     28.994 ±  16.953  ops/s
[info] GraphQLBenchmarks.fragmentsSangriaOld  thrpt    5      0.155 ±   0.039  ops/s
[info] GraphQLBenchmarks.introspectCaliban    thrpt    5   1782.754 ± 309.681  ops/s
[info] GraphQLBenchmarks.introspectSangria    thrpt    5    500.839 ±  15.303  ops/s
[info] GraphQLBenchmarks.simpleCaliban        thrpt    5  26091.734 ± 305.640  ops/s
[info] GraphQLBenchmarks.simpleSangria        thrpt    5   7831.910 ± 139.665  ops/s
```

### CODE REVIEW

File: core/src/main/scala/caliban/execution/Executor.scala [hunk 1]
```diff
@@ -173,28 +172,8 @@ object Executor {
...
 
-  private[caliban] def mergeFields(field: Field, typeName: String): List[Field] = {
-    // ugly mutable code but it's worth it for the speed ;)
-    val array = ArrayBuffer.empty[Field]
-    val map   = collection.mutable.Map.empty[String, Int]
-
-    field.fields.foreach { field =>
-      if (field.condition.forall(_.contains(typeName))) {
-        val name = field.alias.getOrElse(field.name)
-        map.get(name) match {
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/execution/Executor.scala:184]: `map` was never updated 🤦‍♂️

File: core/src/test/scala/caliban/execution/ExecutionSpec.scala [hunk 2]
```diff
@@ -871,6 +870,9 @@ object ExecutionSpec extends DefaultRunnableSpec {
             |    ... on Human {
             |      height
             |    }
+            |    ... on Human {
```
Comment (ghostdogpr) [inline core/src/test/scala/caliban/execution/ExecutionSpec.scala:873]: Added this intentionally to make sure the result has the field only once.

File: core/src/main/scala/caliban/execution/Executor.scala (hunk 1 above)
Comment (frekw) [reply core/src/main/scala/caliban/execution/Executor.scala:184]: Good catch!

**Review by frekw — APPROVED** [review]
I don't know the execution code in detail but I think it looks sensible.

The performance optimizations in the validation also make the code easier to read so a real win-win 🤝

File: core/src/main/scala/caliban/execution/Field.scala [hunk 3]
```diff
@@ -35,7 +36,33 @@ object Field {
     rootType: RootType
   ): Field = {
     def loop(selectionSet: List[Selection], fieldType: __Type): Field = {
-      val fieldList = List.newBuilder[Field]
+      val fieldList  = ArrayBuffer.empty[Field]
```
Comment (paulpdaniels) [inline core/src/main/scala/caliban/execution/Field.scala:39]: Should we give this a size hint as well? I'd expect that to provide a minor improvement in 99% of cases where people don't duplicate their fields (and really only penalizes us in the pathological case).

**Review by paulpdaniels — APPROVED** [review]

File: core/src/main/scala/caliban/execution/Field.scala (hunk 3 above)
Comment (ghostdogpr) [reply core/src/main/scala/caliban/execution/Field.scala:39]: But we don't know the size without going through fragments recursively.

**Pull request merged**